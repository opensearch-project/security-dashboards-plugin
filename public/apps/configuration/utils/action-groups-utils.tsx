/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import { HttpStart } from 'opensearch-dashboards/public';
import { map } from 'lodash';
import { API_ENDPOINT_ACTIONGROUPS, CLUSTER_PERMISSIONS, INDEX_PERMISSIONS } from '../constants';
import { DataObject, ActionGroupItem, ActionGroupUpdate, ObjectsMessage } from '../types';
import { httpDelete, httpGet, httpPost } from './request-utils';
import { getResourceUrl } from './resource-utils';

export interface PermissionListingItem {
  name: string;
  type: 'Action group' | 'Single permission';
  reserved: boolean;
  allowedActions: string[];
  hasClusterPermission: boolean;
  hasIndexPermission: boolean;
}

export async function fetchActionGroups(http: HttpStart): Promise<DataObject<ActionGroupItem>> {
  const actiongroups = await httpGet<ObjectsMessage<ActionGroupItem>>({
    http,
    url: API_ENDPOINT_ACTIONGROUPS,
  });
  return actiongroups.data;
}

export function transformActionGroupsToListingFormat(
  rawData: DataObject<ActionGroupItem>
): PermissionListingItem[] {
  return map(rawData, (value: ActionGroupItem, key?: string) => ({
    name: key || '',
    type: 'Action group',
    reserved: value.reserved,
    allowedActions: value.allowed_actions,
    hasClusterPermission: value.type === 'cluster' || value.type === 'all',
    hasIndexPermission: value.type === 'index' || value.type === 'all',
  }));
}

export async function fetchActionGroupListing(http: HttpStart): Promise<PermissionListingItem[]> {
  return transformActionGroupsToListingFormat(await fetchActionGroups(http));
}

function getClusterSinglePermissions(): PermissionListingItem[] {
  return CLUSTER_PERMISSIONS.map((permission) => ({
    name: permission,
    type: 'Single permission',
    reserved: true,
    allowedActions: [],
    hasClusterPermission: true,
    hasIndexPermission: false,
  }));
}

function getIndexSinglePermissions(): PermissionListingItem[] {
  return INDEX_PERMISSIONS.map((permission) => ({
    name: permission,
    type: 'Single permission',
    reserved: true,
    allowedActions: [],
    hasClusterPermission: false,
    hasIndexPermission: true,
  }));
}

export async function getAllPermissionsListing(http: HttpStart): Promise<PermissionListingItem[]> {
  return mergeAllPermissions(await fetchActionGroups(http));
}

export async function mergeAllPermissions(
  actionGroups: DataObject<ActionGroupItem>
): Promise<PermissionListingItem[]> {
  return transformActionGroupsToListingFormat(actionGroups)
    .concat(getClusterSinglePermissions())
    .concat(getIndexSinglePermissions());
}

export async function updateActionGroup(
  http: HttpStart,
  groupName: string,
  updateObject: ActionGroupUpdate
): Promise<ActionGroupUpdate> {
  return await httpPost({
    http,
    url: getResourceUrl(API_ENDPOINT_ACTIONGROUPS, groupName),
    body: updateObject,
  });
}

export async function requestDeleteActionGroups(http: HttpStart, groups: string[]) {
  for (const group of groups) {
    await httpDelete({ http, url: getResourceUrl(API_ENDPOINT_ACTIONGROUPS, group) });
  }
}
