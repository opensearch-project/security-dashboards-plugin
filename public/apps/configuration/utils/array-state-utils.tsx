/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import { get, set, curry, PropertyName } from 'lodash';
import { Dispatch, SetStateAction } from 'react';

function resolveValue<T>(value: T | (() => T)) {
  return typeof value === 'function' ? (value as Function)() : value;
}

/**
 * Update an element in an sub array
 * @param setStateCallback setState function
 * @param path lodash path, e.g. [0, indexPattern] or "[0].indexPattern"
 * @param newValue value to be updated
 *
 * e.g.
 *  currentState = [{a: 1}, {a: 2}]
 *  path = [0, 'a']
 *  newValue = 10
 * The newState will be [{a: 10}, {a: 2}]
 */
export function updateElementInArray<T>(
  setStateCallback: Dispatch<SetStateAction<any[]>>,
  path: PropertyName | PropertyName[],
  newValue: T | (() => T)
) {
  setStateCallback((prevState) => {
    const newState = [...prevState];
    set(newState, path, resolveValue(newValue));
    return newState;
  });
}

export const updateElementInArrayHandler = curry(updateElementInArray);

/**
 * Append an element to an array (or sub array indicating by path)
 * @param setStateCallback setState function
 * @param path lodash path, e.g. [0, indexPattern] or "[0].indexPattern", use [] to indicate root level
 * @param newValue value to be updated
 *
 * e.g.
 * Scenario 1, path = [] to append to root level array
 *  currentState = [1, 2]
 *  path = []
 *  newValue = 3
 * The newState will be [1, 2, 3]
 *
 * Scenario 2, path != [] to append to sub array
 *  currentState = [{a: [1, 2]}, {a: [3, 4]}]
 *  path = [0, 'a']
 *  newValue = 5
 * The newState will be [{a: [1, 2, 5]}, {a: [3, 4]}]
 */
export function appendElementToArray<T>(
  setStateCallback: Dispatch<SetStateAction<any[]>>,
  path: PropertyName | PropertyName[],
  newValue: T | (() => T)
) {
  const resolvedNewValue = resolveValue(newValue);
  setStateCallback((prevState) => {
    if ((path as PropertyName[]).length === 0) {
      return [...prevState, resolvedNewValue];
    } else {
      const newArray = [...(get(prevState, path) as T[]), resolvedNewValue];
      const newState = [...prevState];
      set(newState, path, newArray);
      return newState;
    }
  });
}

/**
 * Remove an element from an array (or sub array indicating by path)
 * @param setStateCallback setState function
 * @param path lodash path, e.g. [0, indexPattern] or "[0].indexPattern", use [] to indicate root level
 * @param index index of element to be removed
 *
 * e.g.
 * Scenario 1, path = [] to append to root level array
 *  currentState = [1, 2]
 *  path = []
 *  index = 0
 * The newState will be [2]
 *
 * Scenario 2, path != [] to append to sub array
 *  currentState = [{a: [1, 2]}, {a: [3, 4]}]
 *  path = [0, 'a']
 *  index = 1
 * The newState will be [{a: [1]}, {a: [3, 4]}]
 */
export function removeElementFromArray<T>(
  setStateCallback: Dispatch<SetStateAction<any[]>>,
  path: PropertyName | PropertyName[],
  index: number
) {
  setStateCallback((prevState) => {
    if ((path as PropertyName[]).length === 0) {
      const newState = [...prevState];
      newState.splice(index, 1);
      return newState;
    } else {
      const newArray = [...(get(prevState, path) as T[])];
      newArray.splice(index, 1);
      const newState = [...prevState];
      set(newState, path, newArray);
      return newState;
    }
  });
}
