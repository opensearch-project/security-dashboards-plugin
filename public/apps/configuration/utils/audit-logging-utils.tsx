/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import { HttpStart } from 'opensearch-dashboards/public';
import { AuditLoggingSettings } from '../panels/audit-logging/types';
import { API_ENDPOINT_AUDITLOGGING, API_ENDPOINT_AUDITLOGGING_UPDATE } from '../constants';
import { httpGet, httpPost } from './request-utils';

export async function updateAuditLogging(http: HttpStart, updateObject: AuditLoggingSettings) {
  return await httpPost({ http, url: API_ENDPOINT_AUDITLOGGING_UPDATE, body: updateObject });
}

export async function getAuditLogging(http: HttpStart): Promise<AuditLoggingSettings> {
  const rawConfiguration = await httpGet<any>({ http, url: API_ENDPOINT_AUDITLOGGING });
  return rawConfiguration?.config;
}
