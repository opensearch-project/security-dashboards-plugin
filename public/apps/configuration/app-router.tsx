/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import { EuiBreadcrumb, EuiPage, EuiPageBody, EuiPageSideBar } from '@elastic/eui';
import { flow, map, mapValues, partial } from 'lodash';
import React from 'react';
import { HashRouter as Router, Route, Switch, Redirect } from 'react-router-dom';
import { AppDependencies } from '../types';
import { AuditLogging } from './panels/audit-logging/audit-logging';
import { AuditLoggingEditSettings } from './panels/audit-logging/audit-logging-edit-settings';
import {
  SUB_URL_FOR_COMPLIANCE_SETTINGS_EDIT,
  SUB_URL_FOR_GENERAL_SETTINGS_EDIT,
} from './panels/audit-logging/constants';
import { AuthView } from './panels/auth-view/auth-view';
import { GetStarted } from './panels/get-started';
import { InternalUserEdit } from './panels/internal-user-edit/internal-user-edit';
import { NavPanel } from './panels/nav-panel';
import { PermissionList } from './panels/permission-list/permission-list';
import { RoleEdit } from './panels/role-edit/role-edit';
import { RoleList } from './panels/role-list';
import { RoleEditMappedUser } from './panels/role-mapping/role-edit-mapped-user';
import { RoleView } from './panels/role-view/role-view';
import { TenantList } from './panels/tenant-list/tenant-list';
import { UserList } from './panels/user-list';
import { Action, ResourceType, RouteItem, SubAction } from './types';
import { buildHashUrl, buildUrl } from './utils/url-builder';
import { CrossPageToast } from './cross-page-toast';

const LANDING_PAGE_URL = '/getstarted';

const ROUTE_MAP: { [key: string]: RouteItem } = {
  getStarted: {
    name: 'Get Started',
    href: LANDING_PAGE_URL,
  },
  [ResourceType.roles]: {
    name: 'Roles',
    href: buildUrl(ResourceType.roles),
  },
  [ResourceType.users]: {
    name: 'Internal users',
    href: buildUrl(ResourceType.users),
  },
  [ResourceType.permissions]: {
    name: 'Permissions',
    href: buildUrl(ResourceType.permissions),
  },
  [ResourceType.tenants]: {
    name: 'Tenants',
    href: buildUrl(ResourceType.tenants),
  },
  [ResourceType.tenantsManageTab]: {
    name: 'TenantsManageTab',
    href: buildUrl(ResourceType.tenantsManageTab),
  },
  [ResourceType.tenantsConfigureTab]: {
    name: '',
    href: buildUrl(ResourceType.tenantsConfigureTab),
  },
  [ResourceType.auth]: {
    name: 'Authentication',
    href: buildUrl(ResourceType.auth),
  },
  [ResourceType.auditLogging]: {
    name: 'Audit logs',
    href: buildUrl(ResourceType.auditLogging),
  },
};

const ROUTE_LIST = [
  ROUTE_MAP.getStarted,
  ROUTE_MAP[ResourceType.auth],
  ROUTE_MAP[ResourceType.roles],
  ROUTE_MAP[ResourceType.users],
  ROUTE_MAP[ResourceType.permissions],
  ROUTE_MAP[ResourceType.tenants],
  ROUTE_MAP[ResourceType.auditLogging],
  ROUTE_MAP[ResourceType.tenantsConfigureTab],
];

const allNavPanelUrls = ROUTE_LIST.map((route) => route.href).concat([
  buildUrl(ResourceType.auditLogging) + SUB_URL_FOR_GENERAL_SETTINGS_EDIT,
  buildUrl(ResourceType.auditLogging) + SUB_URL_FOR_COMPLIANCE_SETTINGS_EDIT,
]);

export function getBreadcrumbs(
  resourceType?: ResourceType,
  pageTitle?: string,
  subAction?: string
): EuiBreadcrumb[] {
  const breadcrumbs: EuiBreadcrumb[] = [
    {
      text: 'Security',
      href: buildHashUrl(),
    },
  ];

  if (resourceType) {
    breadcrumbs.push({
      text: ROUTE_MAP[resourceType].name,
      href: buildHashUrl(resourceType),
    });
  }

  if (pageTitle) {
    breadcrumbs.push({
      text: pageTitle,
    });
  }

  if (subAction) {
    breadcrumbs.push({
      text: subAction,
    });
  }
  return breadcrumbs;
}

function decodeParams(params: { [k: string]: string }): any {
  return Object.keys(params).reduce((obj: { [k: string]: string }, key: string) => {
    obj[key] = decodeURIComponent(params[key]);
    return obj;
  }, {});
}

export function AppRouter(props: AppDependencies) {
  const setGlobalBreadcrumbs = flow(getBreadcrumbs, props.coreStart.chrome.setBreadcrumbs);

  return (
    <Router basename={props.params.appBasePath}>
      <EuiPage>
        {allNavPanelUrls.map((route) => (
          // Create different routes to update the 'selected' nav item .
          <Route key={route} path={route} exact>
            <EuiPageSideBar>
              <NavPanel items={ROUTE_LIST} />
            </EuiPageSideBar>
          </Route>
        ))}
        <EuiPageBody>
          <Switch>
            <Route
              path={buildUrl(ResourceType.roles, Action.edit) + '/:roleName/' + SubAction.mapuser}
              render={(match) => (
                <RoleEditMappedUser
                  buildBreadcrumbs={partial(setGlobalBreadcrumbs, ResourceType.roles)}
                  {...{ ...props, ...decodeParams(match.match.params) }}
                />
              )}
            />
            <Route
              path={buildUrl(ResourceType.roles, Action.view) + '/:roleName/:prevAction?'}
              render={(match) => (
                <RoleView
                  buildBreadcrumbs={partial(setGlobalBreadcrumbs, ResourceType.roles)}
                  {...{ ...props, ...decodeParams(match.match.params) }}
                />
              )}
            />
            <Route
              path={buildUrl(ResourceType.roles) + '/:action/:sourceRoleName?'}
              render={(match) => (
                <RoleEdit
                  buildBreadcrumbs={partial(setGlobalBreadcrumbs, ResourceType.roles)}
                  {...{ ...props, ...decodeParams(match.match.params) }}
                />
              )}
            />
            <Route
              path={ROUTE_MAP.roles.href}
              render={() => {
                setGlobalBreadcrumbs(ResourceType.roles);
                return <RoleList {...props} />;
              }}
            />
            <Route
              path={ROUTE_MAP.auth.href}
              render={() => {
                setGlobalBreadcrumbs(ResourceType.auth);
                return <AuthView {...props} />;
              }}
            />
            <Route
              path={buildUrl(ResourceType.users) + '/:action/:sourceUserName?'}
              render={(match) => (
                <InternalUserEdit
                  buildBreadcrumbs={partial(setGlobalBreadcrumbs, ResourceType.users)}
                  {...{ ...props, ...decodeParams(match.match.params) }}
                />
              )}
            />
            <Route
              path={ROUTE_MAP.users.href}
              render={() => {
                setGlobalBreadcrumbs(ResourceType.users);
                return <UserList {...props} />;
              }}
            />
            <Route
              path={buildUrl(ResourceType.auditLogging) + SUB_URL_FOR_GENERAL_SETTINGS_EDIT}
              render={() => {
                setGlobalBreadcrumbs(ResourceType.auditLogging, 'General settings');
                return <AuditLoggingEditSettings setting={'general'} {...props} />;
              }}
            />
            <Route
              path={buildUrl(ResourceType.auditLogging) + SUB_URL_FOR_COMPLIANCE_SETTINGS_EDIT}
              render={() => {
                setGlobalBreadcrumbs(ResourceType.auditLogging, 'Compliance settings');
                return <AuditLoggingEditSettings setting={'compliance'} {...props} />;
              }}
            />
            <Route
              path={ROUTE_MAP.auditLogging.href + '/:fromType?'}
              render={(match) => {
                setGlobalBreadcrumbs(ResourceType.auditLogging);
                return <AuditLogging {...{ ...props, ...match.match.params }} />;
              }}
            />
            <Route
              path={ROUTE_MAP.permissions.href}
              render={() => {
                setGlobalBreadcrumbs(ResourceType.permissions);
                return <PermissionList {...props} />;
              }}
            />
            <Route
              path={ROUTE_MAP.tenants.href}
              render={() => {
                setGlobalBreadcrumbs(ResourceType.tenants);
                return <TenantList tabID={'Manage'} {...props} />;
              }}
            />
            <Route
              path={ROUTE_MAP.tenantsConfigureTab.href}
              render={() => {
                setGlobalBreadcrumbs(ResourceType.tenants);
                return <TenantList tabID={'Configure'} {...props} />;
              }}
            />
            <Route
              path={ROUTE_MAP.getStarted.href}
              render={() => {
                setGlobalBreadcrumbs();
                return <GetStarted {...props} />;
              }}
            />
            <Redirect exact from="/" to={LANDING_PAGE_URL} />
          </Switch>
        </EuiPageBody>
        <CrossPageToast />
      </EuiPage>
    </Router>
  );
}
