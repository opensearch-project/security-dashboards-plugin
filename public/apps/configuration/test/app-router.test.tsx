/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import React from 'react';
import { shallow } from 'enzyme';
import { AppRouter } from '../app-router';
import { getDataSourceFromUrl } from '../../../utils/datasource-utils';

jest.mock('../../../utils/datasource-utils', () => ({
  getDataSourceFromUrl: jest.fn(),
}));

describe('SecurityPluginTopNavMenu', () => {
  const coreStartMock = {
    savedObjects: {
      client: jest.fn(),
    },
    notifications: jest.fn(),
    chrome: {
      setBreadcrumbs: jest.fn(),
    },
  };

  const dataSourceMenuMock = jest.fn(() => <div>Mock DataSourceMenu</div>);

  const dataSourceManagementMock = {
    ui: {
      DataSourceMenu: dataSourceMenuMock,
    },
  };

  it('renders DataSourceMenu when dataSource is enabled', () => {
    const securityPluginStartDepsMock = {
      dataSource: {
        dataSourceEnabled: true,
      },
    };

    const wrapper = shallow(
      <AppRouter coreStart={coreStartMock} depsStart={{}} params={{ appBasePath: '' }} />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('getDataSourceFromUrl not called when MDS is disabled', () => {
    const securityPluginStartDepsMock = {
      dataSource: {
        dataSourceEnabled: false,
      },
    };

    shallow(
      <AppRouter
        coreStart={coreStartMock}
        depsStart={securityPluginStartDepsMock}
        params={{ appBasePath: '' }}
      />
    );

    expect(getDataSourceFromUrl).not.toHaveBeenCalled();
  });

  it('getDataSourceFromUrl called when MDS is enabled', () => {
    const securityPluginStartDepsMock = {
      dataSource: {
        dataSourceEnabled: true,
      },
    };

    shallow(
      <AppRouter
        coreStart={coreStartMock}
        depsStart={securityPluginStartDepsMock}
        params={{ appBasePath: '' }}
      />
    );

    expect(getDataSourceFromUrl).toHaveBeenCalled();
  });
});
