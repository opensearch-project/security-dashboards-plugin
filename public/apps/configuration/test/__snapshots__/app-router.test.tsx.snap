// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SecurityPluginTopNavMenu renders DataSourceMenu when dataSource is enabled 1`] = `
<ContextProvider
  value={
    Object {
      "dataSource": Object {
        "id": "",
        "label": "Local cluster",
      },
      "setDataSource": [Function],
    }
  }
>
  <HashRouter
    basename=""
  >
    <EuiPage>
      <Route
        exact={true}
        key="/getstarted"
        path="/getstarted"
      >
        <EuiPageSideBar>
          <NavPanel
            items={
              Array [
                Object {
                  "href": "/getstarted",
                  "name": "Get Started",
                },
                Object {
                  "href": "/auth",
                  "name": "Authentication",
                },
                Object {
                  "href": "/roles",
                  "name": "Roles",
                },
                Object {
                  "href": "/users",
                  "name": "Internal users",
                },
                Object {
                  "href": "/serviceAccounts",
                  "name": "Service Accounts",
                },
                Object {
                  "href": "/permissions",
                  "name": "Permissions",
                },
                Object {
                  "href": "/tenants",
                  "name": "Tenants",
                },
                Object {
                  "href": "/auditLogging",
                  "name": "Audit logs",
                },
                Object {
                  "href": "/tenantsConfigureTab",
                  "name": "",
                },
              ]
            }
          />
        </EuiPageSideBar>
      </Route>
      <Route
        exact={true}
        key="/auth"
        path="/auth"
      >
        <EuiPageSideBar>
          <NavPanel
            items={
              Array [
                Object {
                  "href": "/getstarted",
                  "name": "Get Started",
                },
                Object {
                  "href": "/auth",
                  "name": "Authentication",
                },
                Object {
                  "href": "/roles",
                  "name": "Roles",
                },
                Object {
                  "href": "/users",
                  "name": "Internal users",
                },
                Object {
                  "href": "/serviceAccounts",
                  "name": "Service Accounts",
                },
                Object {
                  "href": "/permissions",
                  "name": "Permissions",
                },
                Object {
                  "href": "/tenants",
                  "name": "Tenants",
                },
                Object {
                  "href": "/auditLogging",
                  "name": "Audit logs",
                },
                Object {
                  "href": "/tenantsConfigureTab",
                  "name": "",
                },
              ]
            }
          />
        </EuiPageSideBar>
      </Route>
      <Route
        exact={true}
        key="/roles"
        path="/roles"
      >
        <EuiPageSideBar>
          <NavPanel
            items={
              Array [
                Object {
                  "href": "/getstarted",
                  "name": "Get Started",
                },
                Object {
                  "href": "/auth",
                  "name": "Authentication",
                },
                Object {
                  "href": "/roles",
                  "name": "Roles",
                },
                Object {
                  "href": "/users",
                  "name": "Internal users",
                },
                Object {
                  "href": "/serviceAccounts",
                  "name": "Service Accounts",
                },
                Object {
                  "href": "/permissions",
                  "name": "Permissions",
                },
                Object {
                  "href": "/tenants",
                  "name": "Tenants",
                },
                Object {
                  "href": "/auditLogging",
                  "name": "Audit logs",
                },
                Object {
                  "href": "/tenantsConfigureTab",
                  "name": "",
                },
              ]
            }
          />
        </EuiPageSideBar>
      </Route>
      <Route
        exact={true}
        key="/users"
        path="/users"
      >
        <EuiPageSideBar>
          <NavPanel
            items={
              Array [
                Object {
                  "href": "/getstarted",
                  "name": "Get Started",
                },
                Object {
                  "href": "/auth",
                  "name": "Authentication",
                },
                Object {
                  "href": "/roles",
                  "name": "Roles",
                },
                Object {
                  "href": "/users",
                  "name": "Internal users",
                },
                Object {
                  "href": "/serviceAccounts",
                  "name": "Service Accounts",
                },
                Object {
                  "href": "/permissions",
                  "name": "Permissions",
                },
                Object {
                  "href": "/tenants",
                  "name": "Tenants",
                },
                Object {
                  "href": "/auditLogging",
                  "name": "Audit logs",
                },
                Object {
                  "href": "/tenantsConfigureTab",
                  "name": "",
                },
              ]
            }
          />
        </EuiPageSideBar>
      </Route>
      <Route
        exact={true}
        key="/serviceAccounts"
        path="/serviceAccounts"
      >
        <EuiPageSideBar>
          <NavPanel
            items={
              Array [
                Object {
                  "href": "/getstarted",
                  "name": "Get Started",
                },
                Object {
                  "href": "/auth",
                  "name": "Authentication",
                },
                Object {
                  "href": "/roles",
                  "name": "Roles",
                },
                Object {
                  "href": "/users",
                  "name": "Internal users",
                },
                Object {
                  "href": "/serviceAccounts",
                  "name": "Service Accounts",
                },
                Object {
                  "href": "/permissions",
                  "name": "Permissions",
                },
                Object {
                  "href": "/tenants",
                  "name": "Tenants",
                },
                Object {
                  "href": "/auditLogging",
                  "name": "Audit logs",
                },
                Object {
                  "href": "/tenantsConfigureTab",
                  "name": "",
                },
              ]
            }
          />
        </EuiPageSideBar>
      </Route>
      <Route
        exact={true}
        key="/permissions"
        path="/permissions"
      >
        <EuiPageSideBar>
          <NavPanel
            items={
              Array [
                Object {
                  "href": "/getstarted",
                  "name": "Get Started",
                },
                Object {
                  "href": "/auth",
                  "name": "Authentication",
                },
                Object {
                  "href": "/roles",
                  "name": "Roles",
                },
                Object {
                  "href": "/users",
                  "name": "Internal users",
                },
                Object {
                  "href": "/serviceAccounts",
                  "name": "Service Accounts",
                },
                Object {
                  "href": "/permissions",
                  "name": "Permissions",
                },
                Object {
                  "href": "/tenants",
                  "name": "Tenants",
                },
                Object {
                  "href": "/auditLogging",
                  "name": "Audit logs",
                },
                Object {
                  "href": "/tenantsConfigureTab",
                  "name": "",
                },
              ]
            }
          />
        </EuiPageSideBar>
      </Route>
      <Route
        exact={true}
        key="/tenants"
        path="/tenants"
      >
        <EuiPageSideBar>
          <NavPanel
            items={
              Array [
                Object {
                  "href": "/getstarted",
                  "name": "Get Started",
                },
                Object {
                  "href": "/auth",
                  "name": "Authentication",
                },
                Object {
                  "href": "/roles",
                  "name": "Roles",
                },
                Object {
                  "href": "/users",
                  "name": "Internal users",
                },
                Object {
                  "href": "/serviceAccounts",
                  "name": "Service Accounts",
                },
                Object {
                  "href": "/permissions",
                  "name": "Permissions",
                },
                Object {
                  "href": "/tenants",
                  "name": "Tenants",
                },
                Object {
                  "href": "/auditLogging",
                  "name": "Audit logs",
                },
                Object {
                  "href": "/tenantsConfigureTab",
                  "name": "",
                },
              ]
            }
          />
        </EuiPageSideBar>
      </Route>
      <Route
        exact={true}
        key="/auditLogging"
        path="/auditLogging"
      >
        <EuiPageSideBar>
          <NavPanel
            items={
              Array [
                Object {
                  "href": "/getstarted",
                  "name": "Get Started",
                },
                Object {
                  "href": "/auth",
                  "name": "Authentication",
                },
                Object {
                  "href": "/roles",
                  "name": "Roles",
                },
                Object {
                  "href": "/users",
                  "name": "Internal users",
                },
                Object {
                  "href": "/serviceAccounts",
                  "name": "Service Accounts",
                },
                Object {
                  "href": "/permissions",
                  "name": "Permissions",
                },
                Object {
                  "href": "/tenants",
                  "name": "Tenants",
                },
                Object {
                  "href": "/auditLogging",
                  "name": "Audit logs",
                },
                Object {
                  "href": "/tenantsConfigureTab",
                  "name": "",
                },
              ]
            }
          />
        </EuiPageSideBar>
      </Route>
      <Route
        exact={true}
        key="/tenantsConfigureTab"
        path="/tenantsConfigureTab"
      >
        <EuiPageSideBar>
          <NavPanel
            items={
              Array [
                Object {
                  "href": "/getstarted",
                  "name": "Get Started",
                },
                Object {
                  "href": "/auth",
                  "name": "Authentication",
                },
                Object {
                  "href": "/roles",
                  "name": "Roles",
                },
                Object {
                  "href": "/users",
                  "name": "Internal users",
                },
                Object {
                  "href": "/serviceAccounts",
                  "name": "Service Accounts",
                },
                Object {
                  "href": "/permissions",
                  "name": "Permissions",
                },
                Object {
                  "href": "/tenants",
                  "name": "Tenants",
                },
                Object {
                  "href": "/auditLogging",
                  "name": "Audit logs",
                },
                Object {
                  "href": "/tenantsConfigureTab",
                  "name": "",
                },
              ]
            }
          />
        </EuiPageSideBar>
      </Route>
      <Route
        exact={true}
        key="/auditLogging/edit/generalSettings"
        path="/auditLogging/edit/generalSettings"
      >
        <EuiPageSideBar>
          <NavPanel
            items={
              Array [
                Object {
                  "href": "/getstarted",
                  "name": "Get Started",
                },
                Object {
                  "href": "/auth",
                  "name": "Authentication",
                },
                Object {
                  "href": "/roles",
                  "name": "Roles",
                },
                Object {
                  "href": "/users",
                  "name": "Internal users",
                },
                Object {
                  "href": "/serviceAccounts",
                  "name": "Service Accounts",
                },
                Object {
                  "href": "/permissions",
                  "name": "Permissions",
                },
                Object {
                  "href": "/tenants",
                  "name": "Tenants",
                },
                Object {
                  "href": "/auditLogging",
                  "name": "Audit logs",
                },
                Object {
                  "href": "/tenantsConfigureTab",
                  "name": "",
                },
              ]
            }
          />
        </EuiPageSideBar>
      </Route>
      <Route
        exact={true}
        key="/auditLogging/edit/complianceSettings"
        path="/auditLogging/edit/complianceSettings"
      >
        <EuiPageSideBar>
          <NavPanel
            items={
              Array [
                Object {
                  "href": "/getstarted",
                  "name": "Get Started",
                },
                Object {
                  "href": "/auth",
                  "name": "Authentication",
                },
                Object {
                  "href": "/roles",
                  "name": "Roles",
                },
                Object {
                  "href": "/users",
                  "name": "Internal users",
                },
                Object {
                  "href": "/serviceAccounts",
                  "name": "Service Accounts",
                },
                Object {
                  "href": "/permissions",
                  "name": "Permissions",
                },
                Object {
                  "href": "/tenants",
                  "name": "Tenants",
                },
                Object {
                  "href": "/auditLogging",
                  "name": "Audit logs",
                },
                Object {
                  "href": "/tenantsConfigureTab",
                  "name": "",
                },
              ]
            }
          />
        </EuiPageSideBar>
      </Route>
      <EuiPageBody>
        <Switch>
          <Route
            path="/roles/edit/:roleName/mapuser"
            render={[Function]}
          />
          <Route
            path="/roles/view/:roleName/:prevAction?"
            render={[Function]}
          />
          <Route
            path="/roles/:action/:sourceRoleName?"
            render={[Function]}
          />
          <Route
            path="/roles"
            render={[Function]}
          />
          <Route
            path="/auth"
            render={[Function]}
          />
          <Route
            path="/users/:action/:sourceUserName?"
            render={[Function]}
          />
          <Route
            path="/users"
            render={[Function]}
          />
          <Route
            path="/serviceAccounts"
            render={[Function]}
          />
          <Route
            path="/auditLogging/edit/generalSettings"
            render={[Function]}
          />
          <Route
            path="/auditLogging/edit/complianceSettings"
            render={[Function]}
          />
          <Route
            path="/auditLogging/:fromType?"
            render={[Function]}
          />
          <Route
            path="/permissions"
            render={[Function]}
          />
          <Route
            path="/tenants"
            render={[Function]}
          />
          <Route
            path="/tenantsConfigureTab"
            render={[Function]}
          />
          <Route
            path="/getstarted"
            render={[Function]}
          />
          <Redirect
            exact={true}
            from="/"
            to="/getstarted"
          />
        </Switch>
      </EuiPageBody>
      <CrossPageToast />
    </EuiPage>
  </HashRouter>
</ContextProvider>
`;
